pkg/apis/storage/validation/validation.go

```diff
@@ -164,8 +164,37 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
 // validateSource tests if the source is valid for VolumeAttachment.
 func validateVolumeAttachmentSource(source *storage.VolumeAttachmentSource, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath, ""))
+	if source.PersistentVolumeName == nil && source.InlineVolumeSource == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, "source", "either persistentVolumeName or inlineVolumeSource must be provided"))
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+		if source.PersistentVolumeName != nil && source.InlineVolumeSource != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, "source", "either persistentVolumeName or inlineVolumeSource can be provided"))
+		}
+	}
+
+	if (source.PersistentVolumeName != nil && len(*source.PersistentVolumeName) == 0) && source.InlineVolumeSource == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, "persistentVolumeName", "persistentVolumeName cannot be empty"))
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
```

Comment:
this is pretty hard to follow... can we gather the `CSIInlineVolume`-protected bits in a single block?

Reply:
yes